### PR TITLE
[FLINK-30217] Replace ListState#clear along with following #add/addAll with a ListState#update

### PR DESCRIPTION
**Pull request by Zakelly**
## What is the purpose of the change

This PR replace ListState#clear along with following ListState#add with a ListState#update, which reduces the state access overhead for KeyedState and normalize the function calls for updating a ListState.

## Brief change log

Found the ListState#clear along with following ListState#add everywhere and did an equivalent replacement for these function calls. I tried my best to do the replacement and there may be omissions.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by Zakelly** [comment]
@masteryhx Would you please take a look? Thanks a lot!

**Review by masteryhx — CHANGES_REQUESTED** [review]
Thanks for the pr.
Existing changes LGTM, besides these, I think we could also update:
1. Doc about operator state example: https://nightlies.apache.org/flink/flink-docs-release-1.17/docs/dev/datastream/fault-tolerance/state/#using-operator-state， see BufferingSink#snapshotState and CounterSource#snapshotState, that many users may learn from.
2. the case of "clear() + addAll()" which could also be replaced.  I saw many examples, e.g. CollectSinkFunction#snapshotState. Of Course, I think we could modify them in a seprate commit.

**Comment by Zakelly** [comment]
@flinkbot run azure

File: flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/SavepointReaderITTestBase.java [hunk 1]
```diff
@@ -293,11 +293,7 @@ public void processBroadcastElement(Integer value, Context ctx, Collector<Void>
...
         @Override
         public void snapshotState(FunctionSnapshotContext context) throws Exception {
-            listState.clear();
-
-            listState.addAll(elements);
-
-            unionState.clear();
+            listState.update(elements);
 
             unionState.addAll(elements);
```
Comment (masteryhx) [inline flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/SavepointReaderITTestBase.java:298]:
```suggestion
            unionState.update(elements);
```

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/sort/StreamSortOperator.java [hunk 2]
```diff
@@ -123,14 +123,12 @@ public void initializeState(StateInitializationContext context) throws Exception
     @Override
     public void snapshotState(StateSnapshotContext context) throws Exception {
         super.snapshotState(context);
-        // clear state first
-        bufferState.clear();
 
         List<Tuple2<RowData, Long>> dataToFlush = new ArrayList<>(inputBuffer.size());
         inputBuffer.forEach((key, value) -> dataToFlush.add(Tuple2.of(key, value)));
 
         // batch put
```
Comment (masteryhx) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/sort/StreamSortOperator.java:130]:
```suggestion
        // batch update
```

**Review by masteryhx — CHANGES_REQUESTED** [review]
Thanks for the update, PTAL my comments.
BTW, `SavepointWriterITCase#StatefulOperator `could also be updated.

**Comment by Zakelly** [comment]
@masteryhx Good catch! Many thanks!

**Review by masteryhx — APPROVED** [review]
Thanks for the update, LGTM.

**Pull request closed without merging**